manta-trusted-setup/src/groth16/phase2.rs

```diff
@@ -0,0 +1,820 @@
+        }
+        if !<C::Pairing as PairingEngineExt>::same_ratio(
+            (ratio_proof.ratio.0, ratio_proof.ratio.1),
+            (prev_state.pk.vk.delta_g2, cur_state.pk.vk.delta_g2),
+        ) {
+            return Err(PhaseTwoError::InconsistentHChange);
+        }
+        if !<C::Pairing as PairingEngineExt>::same_ratio(
+            (prev_state.pk.delta_g1, cur_state.pk.delta_g1),
+            (prev_state.pk.vk.delta_g2, cur_state.pk.vk.delta_g2),
+        ) {
+            return Err(PhaseTwoError::InconsistentHChange);
+        }
+        if !<C::Pairing as PairingEngineExt>::same_ratio(
+            random_linear_combinations::<C>(&cur_state.pk.h_query, &prev_state.pk.h_query),
+            (prev_state.pk.vk.delta_g2, cur_state.pk.vk.delta_g2),
+        ) {
+            return Err(PhaseTwoError::InconsistentHChange);
+        }
+        if !<C::Pairing as PairingEngineExt>::same_ratio(
+            random_linear_combinations::<C>(&cur_state.pk.l_query, &prev_state.pk.l_query),
+            (prev_state.pk.vk.delta_g2, cur_state.pk.vk.delta_g2),
+        ) {
+            return Err(PhaseTwoError::InconsistentLChange);
+        }
+        Ok((
+            cur_state.clone(),
+            Self::challenge(prev_challenge, prev_state, cur_state, ratio_proof),
+        ))
+    }
+
+    #[inline]
+    fn update(
+        state: State<C>,
+        ratio_proof: RatioProof<C>,
+        mut contributions: Contributions<C, N>,
+    ) -> Contributions<C, N> {
+        contributions.states.push(state);
+        contributions.proofs.push(ratio_proof);
+        contributions
+    }
+
+    #[inline]
+    fn verify_transform_all<B, D>(
+        state: State<C>,
+        contributions: Contributions<C, N>,
+        constraint_system: B,
+        accumulator: Accumulator<C>,
+    ) -> Result<(), PhaseTwoError>
+    where
+        B: ConstraintSynthesizer<C::Scalar>,
+        C::Scalar: Sample<D>,
+        C::G1: Sample<D>,
+        C::G2: Sample<D>,
+        State<C>: Clone,
+        RatioProof<C>: Clone,
+        D: Default,
+    {
+        assert_eq!(
+            contributions.proofs.len(),
+            contributions.states.len(),
+            "Langth of `proofs` and `states` does not match."
+        );
+        let (initial_state, initial_contribution) =
+            Self::initialize::<B>(constraint_system, accumulator)?;
+        check_phase_2_invariants(&state, &initial_state)?;
+        if initial_contribution.cs_hash != contributions.cs_hash {
+            return Err(PhaseTwoError::ConstraintSystemHashesDiffer);
+        }
+        let mut challenge = initial_contribution.cs_hash;
+        let mut prev_state = initial_state;
+        for (proof, state) in contributions.proofs.iter().zip(&contributions.states) {
+            (prev_state, challenge) =
+                Self::verify_transform::<D>(challenge, prev_state, state.clone(), proof.clone())
+                    .expect("Verify transform failed.");
```

Comment:
There's a cheaper check that goes like this:
```
// Check phase 2 invariants
let mut current_delta = G1::one();
for (proof, state) in contributions {
   // Check the ratio proof is valid
   // Check the ratio `(current_delta, state.pk.delta_g1)` matches the ratio in `proof`
   current_delta = state.pk.delta_g1;
}
// Check `current_delta == state.pk.delta_g1`
// Check that ratio (G1::one(), state.pk.delta_g1) equals ratio (G2::one(), state.pk.delta_g2)
// Check that `h_query` and `l_query` were correctly updated
```
This works because the only change in phase 2 is that some terms of the proving key are multiplied by an overall factor `delta`.  That factor `delta` is composite, the product of many factors.  We don't have a knowledge proof for the overall `delta`, but we can effectively make one by chaining together the knowledge proofs for each `factor`.  That's what's happening in the `for` loop.  Finally, you just check that the various pieces of the _final_ proving key have been updated correctly.  This implies that they must have been updated correctly in each intermediate step.